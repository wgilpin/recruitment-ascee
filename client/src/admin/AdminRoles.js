import React from 'react';
import PropTypes from 'prop-types';
import { TabPanel } from 'react-tabs';
import Alt from '../Alt';
import UpImg from '../images/arrow_up_24x24.png';
import DownImg from '../images/arrow_down_24x24.png';
import TableStyles from '../TableStyles';
import SearchImg from '../images/magnifying_glass_24x24.png';
import FetchData from '../common/FetchData';


const primary = TableStyles.styles.themeColor.color;

const styles = {
  searchButton: {
    borderRadius: '3px',
    borderStyle: 'solid',
    borderWidth: '1px',
    borderColor: 'darkgray',
    backgroundColor: '#0000',
    padding: '6px',
  },
  smallButtonImg: {
    width: '20px',
    height: '20px',
  },
  h1: {
    color: primary,
  },
  h2: {
    color: primary,
  },
  userLine: {
    justifyItems: 'left',
    textAlign: 'left',
    padding: '12px',
  },
  moveButtons: {
    paddingLeft: '24px',
    position: 'relative',
    top: '11px',
  },
  searchInput: {
    backgroundColor: `#000`,
    borderStyle: `solid`,
    borderWidth: `1px`,
    borderColor: `darkgray`,
    height: `23px`,
    padding: `6px`,
    float: 'right',
  },
  searchBtnOuter: {
    float: 'right',
    cursor: 'pointer',
    marginRight: '12px',
  },
  tab: {
    width: '20%',
  }
};

const propTypes = {};

const defaultProps = {};

export default class AdminRoles extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      staff: {},
      searchText: '',
      searchResults: [],
    };
  }

  arrayToObject(arr, keyField) {
    return arr.reduce((acc, val) => ({ ...acc, [val[keyField]]: val}), {})
  }

  componentWillMount() {
    new FetchData({ scope: 'admin/users' })
      .get()
      .then(data => this.setState({ staff: this.arrayToObject(data.info || {}, 'id') }));
  }

  filterStaff(role) {
    // filter by is_<role>, e.g. is_admin
    return Object
      .values(this.state.staff || {})
      .filter(c => !!(c[`is_${role}`]));
  }

  handleSearch() {
    // TODO:
    new FetchData({ scope: 'character/find', param1: this.state.searchText })
      .get()
      .then(res => {
        this.setState({ searchResults: res });
      });
  }

  async writeRoles(id, newRole) {
    let roles = {
      recruiter: false,
      senior_recruiter: false,
      admin: false,
    }
    switch(newRole) {
      case 1:
        roles.recruiter = true;
        break;
      case 2:
        roles.senior_recruiter = true;
        break;
      case 3:
        roles.admin = true;
        break;
      default:
        // nada
    }
    await new FetchData({ id, scope: 'admin', param1: 'set_roles' })
      .get(roles);
    return new FetchData({ scope: 'admin/users' })
      .get()
      .then(data => this.setState({ staff: data }));
  }

  handleMove(id, direction) {
    // TODO: + confirm dixt in API
    const user = this.state.staff[id];
    let oldState = 0;
    if (user.isRecruiter) {
      oldState = 1;
    }
    if (user.isSnrRecruiter) {
      oldState = 2;
    }
    if (user.isAdmin) {
      oldState = 3;
    }
    let newState = oldState + (direction === 'up' ? 1 : -1);
    newState = Math.min(newState, 3);
    newState = Math.max(newState, 0);
    this.writeRoles(id, newState).then(() => {
      this.setState({
        staff: {
          ...this.state.staff,
          [id]: {
            ...this.state.staff[id],
            isAdmin: newState === 3,
            isSnrRecruiter: newState === 2,
            isRecruiter: newState === 1,
          },
        },
      });
    } )

  }

  makeSearchResultLine(char) {
    // TODO:
    return (
      <div>
        {char.name}
        <img
          style={styles.moveButtons}
          src={UpImg}
          alt="up"
          onClick={() => this.handleMove(char.id, 'up')}
        />
        <img
          style={styles.moveButtons}
          src={DownImg}
          alt="up"
          onClick={() => this.handleMove(char.id, 'down')}
        />
      </div>
    );
  }

  sortByNameFn(a, b) {
    if (a.name > b.name) {
      return 1;
    }
    if (a.name < b.name) {
      return -1;
    }
    return 0;
  }

  sectionList(label, list) {
    /*
     * create a list of users for display
     *
     * @param {label} string - text lable for the section
     * @param {list} [object] - list of users
     * @returns jsx
     */
    return [
      <div style={styles.h2}>{label}</div>,
      list.length > 0 ? (
        list
          .sort(this.sortByNameFn)
          .map(user => this.recruitLine(user.id, user))
      ) : (
          <div style={styles.noneText}>None</div>
        ),
    ];
  }

  recruitLine(id, recruiter) {
    return (
      <div key={id} style={styles.userLine}>
        <Alt
          onClick={() => this.handleClick(id)}
          name={recruiter.name}
          id={recruiter.id}
          style={{ float: 'left', width: '250px' }} />
        <img
          style={styles.moveButtons}
          src={UpImg}
          alt="up"
          onClick={() => this.handleMove(id, 'up')}
        />
        <img
          style={styles.moveButtons}
          src={DownImg}
          alt="up"
          onClick={() => this.handleMove(id, 'down')}
        />
      </div>
    );
  }

  updateSearchText(evt) {
    this.setState({
      searchText: evt.target.value,
    });
  }

  buildRolesPanel = () => {
    const admins = this.filterStaff('admin');
    const recruiters = this.filterStaff('recruiter');
    const snrRecruiters = this.filterStaff('snr_recruiter');
    return <>
        <h2>Roles</h2>
        {this.sectionList('Admins', admins)}
        <hr />
        {this.sectionList('Senior Recruiters', snrRecruiters)}
        <hr />
        {this.sectionList('Recruiters', recruiters)}
        <hr />
        <div style={styles.h2}>Others</div>
        <div>
          <div>
            <input
              style={styles.searchInput}
              type="text"
              placeholder="search..."
              onChange={this.updateSearchText}
              value={this.state.searchText}
            />
          </div>
          <div>
            {this.state.searchResults.forEach(char =>
              this.makeSearchResultLine(char),
            )}
          </div>
          <div style={styles.searchBtnOuter}>
            <button onClick={this.handleSearch} style={styles.searchButton}>
              <img style={styles.smallButtonImg} src={SearchImg} alt="Search" />
            </button>
          </div>
        </div>
      </>

  };

  render() {
    return this.buildRolesPanel();
  }
}

AdminRoles.propTypes = propTypes;
AdminRoles.defaultProps = defaultProps;
