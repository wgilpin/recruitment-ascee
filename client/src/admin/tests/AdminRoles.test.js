import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import AdminRoles from './../AdminRoles';

let mockFetch;

beforeEach(() => {
  const serverState = [
    {
      id: 1234,
      name: 'user 1',
      is_recruiter: true,
      is_senior_recruiter: false,
      is_admin: false,
    },
    {
      id: 4321,
      name: 'user 2',
      is_recruiter: true,
      is_senior_recruiter: true,
      is_admin: false,
    },
    {
      id: 1289,
      name: 'user 3',
      is_recruiter: true,
      is_senior_recruiter: true,
      is_admin: true,
    },
    {
      id: 8921,
      name: 'user 4',
      is_recruiter: false,
      is_senior_recruiter: false,
      is_admin: true,
    },
  ];

  mockFetch = jest.fn().mockImplementation((scope, params) => {
    
    if (scope.scope === 'admin/users') {
      
      return Promise.resolve({
        info: serverState,
      });
    } else if (scope.param1==='set_roles') {
      
      const index = serverState.findIndex(item => item.id === scope.id);
      serverState[index] = params;
    } else {
      throw new Error("Scope Not Valid");
    }
  });
});

describe('<AdminRoles>', () => {
  it('renders without crashing', () => {
    shallow(<AdminRoles />);
  });

  it('loads', async done => {
    var wrapper = mount(<AdminRoles fetcher={mockFetch} />);
    setImmediate(() => {
      const { staff, showConfirm } = wrapper.state();
      expect(showConfirm).toEqual(false);
      expect(Object.keys(staff).sort()).toEqual(
        ['1234', '1289', '4321', '8921'].sort()
      );
      done();
    });
  });

  const toggleStatusHelper = async (field, done) => {
    var wrapper = mount(<AdminRoles fetcher={mockFetch} />);
    await Promise.resolve();
    
    wrapper.update();
    const currentState = wrapper.state().staff[1234][field];
    wrapper.instance().handleClickCheck(1234, field);
    wrapper.instance().doChange();
    await Promise.resolve();
    await Promise.resolve();
    
    const newState = wrapper.state().staff[1234][field];
    
    expect(currentState).not.toEqual(newState);
    done();
  };

  it('allows toggling recruiter', async done => {
    toggleStatusHelper('is_recruiter', done);
  });

  it('allows toggling senior recruiter', async done => {
    toggleStatusHelper('is_senior_recruiter', done);
  });
  
  it('allows toggling admin', async done => {
    toggleStatusHelper('is_admin', done);
  });

  it('matches snapshot', async () => {
    const tree = renderer.create(<AdminRoles fetcher={mockFetch} />);
    await Promise.resolve();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
