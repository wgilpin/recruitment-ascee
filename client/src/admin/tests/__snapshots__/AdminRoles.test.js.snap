// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminRoles> matches snapshot 1`] = `
Array [
  <div
    style={
      Object {
        "display": "table",
        "marginLeft": "auto",
        "marginRight": "auto",
        "maxWidth": "500px",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "color": "#01799A",
          "display": "table-header-group",
          "fontWeight": "bold",
          "paddingBottom": "19px",
        }
      }
    >
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "paddingRight": "6px",
            "textAlign": "left",
          }
        }
      >
        Name
      </div>
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "paddingRight": "6px",
            "textAlign": "left",
          }
        }
      >
        Recruiter
      </div>
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "paddingRight": "6px",
            "textAlign": "left",
          }
        }
      >
        Senior
      </div>
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "paddingRight": "6px",
            "textAlign": "left",
          }
        }
      >
        Admin
      </div>
    </div>
    <div
      id="roles-list"
      style={
        Object {
          "display": "table-row-group",
        }
      }
    >
      <div
        style={
          Object {
            "display": "table-row",
          }
        }
      >
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <div
            onClick={[Function]}
            style={
              Object {
                "cursor": null,
                "display": "table",
                "padding": 8,
              }
            }
          >
            <div
              style={
                Object {
                  "display": "table-row",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "table-cell",
                    "paddingLeft": "6px",
                    "paddingRight": "6px",
                    "textAlign": "left",
                  }
                }
              >
                <img
                  alt="Avatar"
                  onClick={[Function]}
                  src="https://image.eveonline.com/Character/1234_64.jpg"
                  style={
                    Object {
                      "backgroundColor": "black",
                      "borderRadius": "50%",
                      "boxShadow": "1px 3px 1px #111",
                      "height": 32,
                      "width": 32,
                    }
                  }
                />
              </div>
              <div
                style={
                  Object {
                    "display": "table-cell",
                    "height": "24px",
                    "marginBottom": "auto",
                    "marginTop": "auto",
                    "paddingLeft": "8px",
                    "paddingRight": "6px",
                    "size": 14,
                    "textAlign": "left",
                    "verticalAlign": "middle",
                  }
                }
              >
                user 1
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_recruiter"
            onClick={[Function]}
            src="check_box_white.png"
          />
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_senior_recruiter"
            onClick={[Function]}
            src="check_box_outline_blank.png"
          />
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_admin"
            onClick={[Function]}
            src="check_box_outline_blank.png"
          />
        </div>
      </div>
      <div
        style={
          Object {
            "display": "table-row",
          }
        }
      >
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <div
            onClick={[Function]}
            style={
              Object {
                "cursor": null,
                "display": "table",
                "padding": 8,
              }
            }
          >
            <div
              style={
                Object {
                  "display": "table-row",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "table-cell",
                    "paddingLeft": "6px",
                    "paddingRight": "6px",
                    "textAlign": "left",
                  }
                }
              >
                <img
                  alt="Avatar"
                  onClick={[Function]}
                  src="https://image.eveonline.com/Character/4321_64.jpg"
                  style={
                    Object {
                      "backgroundColor": "black",
                      "borderRadius": "50%",
                      "boxShadow": "1px 3px 1px #111",
                      "height": 32,
                      "width": 32,
                    }
                  }
                />
              </div>
              <div
                style={
                  Object {
                    "display": "table-cell",
                    "height": "24px",
                    "marginBottom": "auto",
                    "marginTop": "auto",
                    "paddingLeft": "8px",
                    "paddingRight": "6px",
                    "size": 14,
                    "textAlign": "left",
                    "verticalAlign": "middle",
                  }
                }
              >
                user 2
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_recruiter"
            onClick={[Function]}
            src="check_box_white.png"
          />
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_senior_recruiter"
            onClick={[Function]}
            src="check_box_white.png"
          />
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_admin"
            onClick={[Function]}
            src="check_box_outline_blank.png"
          />
        </div>
      </div>
      <div
        style={
          Object {
            "display": "table-row",
          }
        }
      >
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <div
            onClick={[Function]}
            style={
              Object {
                "cursor": null,
                "display": "table",
                "padding": 8,
              }
            }
          >
            <div
              style={
                Object {
                  "display": "table-row",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "table-cell",
                    "paddingLeft": "6px",
                    "paddingRight": "6px",
                    "textAlign": "left",
                  }
                }
              >
                <img
                  alt="Avatar"
                  onClick={[Function]}
                  src="https://image.eveonline.com/Character/1289_64.jpg"
                  style={
                    Object {
                      "backgroundColor": "black",
                      "borderRadius": "50%",
                      "boxShadow": "1px 3px 1px #111",
                      "height": 32,
                      "width": 32,
                    }
                  }
                />
              </div>
              <div
                style={
                  Object {
                    "display": "table-cell",
                    "height": "24px",
                    "marginBottom": "auto",
                    "marginTop": "auto",
                    "paddingLeft": "8px",
                    "paddingRight": "6px",
                    "size": 14,
                    "textAlign": "left",
                    "verticalAlign": "middle",
                  }
                }
              >
                user 3
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_recruiter"
            onClick={[Function]}
            src="check_box_white.png"
          />
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_senior_recruiter"
            onClick={[Function]}
            src="check_box_white.png"
          />
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_admin"
            onClick={[Function]}
            src="check_box_white.png"
          />
        </div>
      </div>
      <div
        style={
          Object {
            "display": "table-row",
          }
        }
      >
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <div
            onClick={[Function]}
            style={
              Object {
                "cursor": null,
                "display": "table",
                "padding": 8,
              }
            }
          >
            <div
              style={
                Object {
                  "display": "table-row",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "table-cell",
                    "paddingLeft": "6px",
                    "paddingRight": "6px",
                    "textAlign": "left",
                  }
                }
              >
                <img
                  alt="Avatar"
                  onClick={[Function]}
                  src="https://image.eveonline.com/Character/8921_64.jpg"
                  style={
                    Object {
                      "backgroundColor": "black",
                      "borderRadius": "50%",
                      "boxShadow": "1px 3px 1px #111",
                      "height": 32,
                      "width": 32,
                    }
                  }
                />
              </div>
              <div
                style={
                  Object {
                    "display": "table-cell",
                    "height": "24px",
                    "marginBottom": "auto",
                    "marginTop": "auto",
                    "paddingLeft": "8px",
                    "paddingRight": "6px",
                    "size": 14,
                    "textAlign": "left",
                    "verticalAlign": "middle",
                  }
                }
              >
                user 4
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_recruiter"
            onClick={[Function]}
            src="check_box_outline_blank.png"
          />
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_senior_recruiter"
            onClick={[Function]}
            src="check_box_outline_blank.png"
          />
        </div>
        <div
          style={
            Object {
              "display": "table-cell",
              "paddingLeft": "6px",
              "paddingRight": "6px",
              "textAlign": "left",
            }
          }
        >
          <img
            alt="is_admin"
            onClick={[Function]}
            src="check_box_white.png"
          />
        </div>
      </div>
    </div>
  </div>,
  <div
    style={
      Object {
        "marginBottom": "24px",
        "marginTop": "24px",
      }
    }
  >
    <div
      style={
        Object {
          "marginLeft": "auto",
          "marginRight": "auto",
          "width": "50%",
        }
      }
    >
      <div
        style={
          Object {
            "display": "inline-block",
            "padding": "6px",
          }
        }
      >
        <input
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="search..."
          style={
            Object {
              "backgroundColor": "black",
              "borderBottom": "1px white solid",
              "borderColor": "darkgray",
              "borderLeft": "black",
              "borderRight": "black",
              "borderStyle": "solid",
              "borderTop": "black",
              "borderWidth": "1px",
              "color": "white",
              "padding": "6px",
            }
          }
          type="text"
        />
      </div>
      <div
        style={
          Object {
            "cursor": "pointer",
            "display": "inline-block",
            "marginRight": "12px",
          }
        }
      >
        <button
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#0000",
              "borderStyle": "none",
              "padding": "6px",
            }
          }
        >
          <img
            alt="Search"
            src="magnifying_glass_24x24.png"
            style={
              Object {
                "height": "20px",
                "width": "20px",
              }
            }
          />
        </button>
      </div>
    </div>
  </div>,
]
`;
